### PR TITLE
Better interfaces (task #809)

### PR DESCRIPTION
**Pull request by softius**
This part of the work required for task #809 

* Add MenuItemContainerInterface to consolidate methods between MenuItemInterface and MenuInterface
* Provide a default implementation (trait)  for MenuItemContainerInterface
* Define common setters and getters under MenuItemInterface
* Improve type hinting for MenuItemInterface implementations  …
* Deprecate MenuInterface::addChild, MenuInterface::getChildren and MenuInterface::removeChild
* Add tests for removeMenuItem method

### CODE REVIEW

File: src/MenuBuilder/BaseMenuRenderClass.php [hunk 1]
```diff
@@ -136,40 +148,22 @@ protected function renderMenuItem($item)
...
-                break;
-            case 'Menu\MenuBuilder\MenuItemLinkButtonModal':
-                $result = $this->_buildLinkButtonModal($item, $extLabel);
-                break;
-            case 'Menu\MenuBuilder\MenuItemLinkModal':
-                $result = $this->_buildLinkModal($item, $extLabel);
-                break;
-            default:
-                $result = $this->_buildLink($item, $extLabel);
+        if (array_key_exists($class, self::$itemRenderer)) {
```
Comment (mamchenkov) [inline src/MenuBuilder/BaseMenuRenderClass.php]: @softius I think we can further improve this, by relying on the naming convention.

File: src/MenuBuilder/BaseMenuItem.php [hunk 2]
```diff
@@ -306,36 +303,38 @@ public function setDataType($dataType)
      *  addChild method
      *
      * @param MenuItemInterface $child menu item
+     * @deprecated Use add method instead.
      * @return void
      */
     public function addChild(MenuItemInterface $child)
     {
-        array_push($this->children, $child);
+        trigger_error('Method ' . __METHOD__ . ' is deprecated. Use add method instead', E_USER_DEPRECATED);
```
Comment (georgeconstantinou) [inline src/MenuBuilder/BaseMenuItem.php]: @softius I believe `addMenuItem` is the correct method name for the deprecated error message.

File: src/MenuBuilder/BaseMenuItem.php [hunk 3]
```diff
@@ -306,36 +303,38 @@ public function setDataType($dataType)
...
      * removeChild method
      *
      * @param string $childId to be removed
+     * @deprecated Use remove method instead.
      * @return void
      */
     public function removeChild($childId)
     {
-        throw new NotImplementedException('Method ' . __METHOD__ . ' is not implemented yet!');
+        trigger_error('Method ' . __METHOD__ . ' is deprecated. Use remove method instead', E_USER_DEPRECATED);
```
Comment (georgeconstantinou) [inline src/MenuBuilder/BaseMenuItem.php]: @softius same here, `removeMenuItem` is the correct method name.

File: src/MenuBuilder/BaseMenuItem.php [hunk 4]
```diff
@@ -306,36 +303,38 @@ public function setDataType($dataType)
...
      *
+     * @deprecated Use getAll method instead.
      * @return array list of child items
      */
     public function getChildren()
     {
-        usort($this->children, function ($a, $b) {
-            return $a->getOrder() > $b->getOrder();
-        });
+        trigger_error('Method ' . __METHOD__ . ' is deprecated. Use getAll method instead', E_USER_DEPRECATED);
```
Comment (georgeconstantinou) [inline src/MenuBuilder/BaseMenuItem.php]: @softius same here, `getMenuItems` is the correct name.

File: src/MenuBuilder/BaseMenuItem.php (hunk 2 above)
Comment (softius) [reply src/MenuBuilder/BaseMenuItem.php]: Nice catch! :+1:

**Pull request merged**